cw_solana/lib/solana_client.dart

```diff
@@ -468,108 +627,170 @@ class SolanaWalletClient {
     return pendingTransaction;
   }
 
+  SolanaTransaction _constructNativeTransaction({
+    required SolanaPrivateKey ownerPrivateKey,
+    required String destinationAddress,
+    required SolAddress latestBlockhash,
+    required int lamports,
+  }) {
+    final owner = ownerPrivateKey.publicKey().toAddress();
+
+    /// Create a transfer instruction to move funds from the owner to the receiver.
+    final transferInstruction = SystemProgram.transfer(
+      from: owner,
+      layout: SystemTransferLayout(lamports: BigInt.from(lamports)),
+      to: SolAddress(destinationAddress),
+    );
+
+    /// Construct a Solana transaction with the transfer instruction.
+    return SolanaTransaction(
+      instructions: [transferInstruction],
+      recentBlockhash: latestBlockhash,
+      payerKey: ownerPrivateKey.publicKey().toAddress(),
+      type: TransactionType.v0,
+    );
+  }
+
+  Future<ProgramDerivedAddress?> _getOrCreateAssociatedTokenAccount({
+    required SolanaPrivateKey payerPrivateKey,
+    required SolAddress ownerAddress,
+    required SolAddress mintAddress,
+    required bool shouldCreateATA,
+  }) async {
+    final associatedTokenAccount = AssociatedTokenAccountProgramUtils.associatedTokenAccount(
+      mint: mintAddress,
+      owner: ownerAddress,
+    );
+
+    SolanaAccountInfo? accountInfo;
+    try {
+      accountInfo = await _provider!.request(
+        SolanaRPCGetAccountInfo(account: associatedTokenAccount.address),
+      );
+    } catch (e) {
+      accountInfo = null;
+    }
+
+    // If aacountInfo is null, signifies that the associatedTokenAccount has only been created locally and not been broadcasted to the blockchain.
+    if (accountInfo != null) return associatedTokenAccount;
+
+    if (!shouldCreateATA) return null;
+
+    final createAssociatedTokenAccount = AssociatedTokenAccountProgram.associatedTokenAccount(
+      payer: payerPrivateKey.publicKey().toAddress(),
+      associatedToken: associatedTokenAccount.address,
+      owner: ownerAddress,
+      mint: mintAddress,
+    );
+
+    final blockhash = await _getLatestBlockhash(Commitment.confirmed);
+
+    final transaction = SolanaTransaction(
+      payerKey: payerPrivateKey.publicKey().toAddress(),
+      instructions: [createAssociatedTokenAccount],
+      recentBlockhash: blockhash,
+    );
+
+    transaction.sign([payerPrivateKey]);
```

Comment:
why this is different than this `ownerPrivateKey.sign(transaction.serializeMessage());`?

Reply:
It's almost the same thing/flow that they use beneath the surface

The additional thing this does is to add the signature gotten from signing with the private keys to an internal list of signature belonging to that transaction, before it's sent it to the chain.